File: app/src/main/java/com/asmat/rolando/popularmovies/ui/fragments/WatchLaterMoviesGridFragment.java
package com.asmat.rolando.popularmovies.ui.fragments;

import android.arch.lifecycle.LiveData;
import com.asmat.rolando.popularmovies.ui.adapters.grid.MoviesGridBaseAdapter;
import com.asmat.rolando.popularmovies.ui.adapters.grid.WatchLaterMoviesGridAdapter;
import com.asmat.rolando.popularmovies.database.DatabaseManager;
import com.asmat.rolando.popularmovies.ui.adapters.MovieAdapterOnClickHandler;

public class WatchLaterMoviesGridFragment extends BaseGridFragment {

    @Override
    MoviesGridBaseAdapter getAdapter(MovieAdapterOnClickHandler handler) {
        return new WatchLaterMoviesGridAdapter(handler);
    }

    @Override
    LiveData<Movie[]> getMovieSource() {
        return DatabaseManager.INSTANCE.getWatchLaterMovies();
    }
}

File: app/src/main/java/com/asmat/rolando/popularmovies/ui/fragments/WatchLaterMoviesGridFragment.kt
package com.asmat.rolando.popularmovies.ui.fragments

import androidx.lifecycle.LiveData
import androidx.lifecycle.Transformations
import com.asmat.rolando.popularmovies.ui.mylists.WatchLaterMoviesGridAdapter
import com.asmat.rolando.popularmovies.model.Movie
import com.asmat.rolando.popularmovies.model.mappers.MovieMapper
import com.asmat.rolando.popularmovies.ui.common.MovieAdapterOnClickHandler
import com.asmat.rolando.popularmovies.ui.common.BaseGridFragment
import com.asmat.rolando.popularmovies.ui.common.BaseMoviesGridAdapter

class WatchLaterMoviesGridFragment : BaseGridFragment() {

    override val movieSource: LiveData<List<Movie>>
        get() {
            val source = moviesRepository.getAllWatchLaterMovies()
            return Transformations.map(source) { watchLaterMovies ->
                watchLaterMovies.map {
                    MovieMapper.from(it)
                }
            }
        }

    override fun getAdapter(handler: MovieAdapterOnClickHandler): BaseMoviesGridAdapter {
        return WatchLaterMoviesGridAdapter(handler)
    }
}